x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -1587,4 +1589,72 @@ private static LogicalPlan normalize(Aggregate aggregate, AttributeMap<Expressio
             return changed.get() ? new Aggregate(aggregate.source(), aggregate.child(), aggregate.groupings(), newAggs) : aggregate;
         }
     }
+
+    public static class FoldNull extends OptimizerRules.OptimizerExpressionRule<Expression> {
+
+        public FoldNull() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected Expression rule(Expression e) {
+            if (e instanceof In in) {
+                if (Expressions.isNull(in.value())) {
+                    return Literal.of(in, null);
+                }
+            } else if (e instanceof Alias == false
+                && e.nullable() == Nullability.TRUE
+                && Expressions.anyMatch(e.children(), Expressions::isNull)) {
+                    return Literal.of(e, null);
+                }
+            return e;
+        }
+    }
+
+    public static class PropagateNullable extends OptimizerRules.PropagateNullable {
+        @Override
+        protected Expression rule(And and) {
+            List<Expression> splits = Predicates.splitAnd(and);
+
+            Set<Expression> nullExpressions = new LinkedHashSet<>();
+            Set<Expression> notNullExpressions = new LinkedHashSet<>();
+            List<Expression> others = new LinkedList<>();
+
+            // first find isNull/isNotNull
+            for (Expression ex : splits) {
+                // Avoid considering constants. It would lead to collisions with foldable propagation
+                // Folding will take care of them eventually anyway
+                if (ex instanceof IsNull isn && isn.field().foldable() == false) {
+                    nullExpressions.add(isn.field());
+                } else if (ex instanceof IsNotNull isnn && isnn.field().foldable() == false) {
+                    notNullExpressions.add(isnn.field());
+                }
+                // the rest
+                else {
+                    others.add(ex);
+                }
+            }
+
+            // check for is isNull and isNotNull --> FALSE
+            if (Sets.haveNonEmptyIntersection(nullExpressions, notNullExpressions)) {
+                return Literal.of(and, Boolean.FALSE);
+            }
+
+            // apply nullability across relevant/matching expressions
+
+            // first against all nullable expressions
+            // followed by all not-nullable expressions
+            boolean modified = replace(nullExpressions, others, splits, this::nullify);
+            modified |= replace(notNullExpressions, others, splits, this::nonNullify);
+            if (modified) {
+                // reconstruct the expression
+                return Predicates.combineAnd(splits);
+            }
+            return and;
+        }
+
+        protected Expression nullify(Expression exp, Expression nullExp) {
+            return Literal.of(exp, null);
+        }
```

Comment:
This doesn't look right - to always return null regardless of the nullExp.

Reply:
I see what you mean, it seems a stretch, but:
- the parent method returns null regardless of the nullExp as well
- I cannot speculate on expression nullability, because we know in ESQL the nullability logic is inconsistent

SQL only uses `nullExp` to do one additional optimization on `coalesce()` (remove the expressions that we know are null); I'll add this optimization, but it doesn't change the correctness of the rule.